client/network/src/bitswap.rs

```diff
@@ -0,0 +1,337 @@
+		let mut buf = varint_encode::u64_buffer();
+		let mh_len = varint_encode::u64(self.mh_len as u64, &mut buf);
+		res.extend_from_slice(mh_len);
+		res
+	}
+}
+
+/// Network behaviour that handles sending and receiving IPFS blocks.
+pub struct Bitswap<B> {
+	client: Arc<dyn Client<B>>,
+	ready_blocks: VecDeque<(PeerId, BitswapMessage)>,
+}
+
+impl<B: BlockT> Bitswap<B> {
+	/// Create a new instance of the bitswap protocol handler.
+	pub fn new(client: Arc<dyn Client<B>>) -> Self {
+		Bitswap {
+			client,
+			ready_blocks: Default::default(),
+		}
+	}
+}
+
+impl<B: BlockT> NetworkBehaviour for Bitswap<B> {
+	type ProtocolsHandler = OneShotHandler<BitswapConfig, BitswapMessage, HandlerEvent>;
+	type OutEvent = void::Void;
+
+	fn new_handler(&mut self) -> Self::ProtocolsHandler {
+		Default::default()
+	}
+
+	fn addresses_of_peer(&mut self, _peer: &PeerId) -> Vec<Multiaddr> {
+		Vec::new()
+	}
+
+	fn inject_connected(&mut self, _peer: &PeerId) {
+	}
+
+	fn inject_disconnected(&mut self, _peer: &PeerId) {
+	}
+
+	fn inject_event(&mut self, peer: PeerId, _connection: ConnectionId, message: HandlerEvent) {
+		let request = match message {
+			HandlerEvent::ResponseSent => return,
+			HandlerEvent::Request(msg) => msg,
+		};
+		trace!(target: LOG_TARGET, "Received request: {:?} from {}", request, peer);
+		if self.ready_blocks.len() > MAX_RESPONSE_QUEUE {
+			debug!(target: LOG_TARGET, "Ignored request: queue is full");
+			return;
+		}
+		let mut response = BitswapMessage {
+			wantlist: None,
+			blocks: Default::default(),
+			payload: Default::default(),
+			block_presences: Default::default(),
+			pending_bytes: 0,
+		};
+		let wantlist = match request.wantlist {
+			Some(wantlist) => wantlist,
+			None => {
+				debug!(
+					target: LOG_TARGET,
+					"Unexpected bitswap message from {}",
+					peer,
+				);
+				return;
+			}
+		};
+		if wantlist.entries.len() > MAX_WANTED_BLOCKS {
+			trace!(target: LOG_TARGET, "Ignored request: too many entries");
+			return;
+		}
+		for entry in wantlist.entries {
+			let cid = match cid::Cid::read_bytes(entry.block.as_slice()) {
+				Ok(cid) => cid,
+				Err(e) => {
+					trace!(target: LOG_TARGET, "Bad CID {:?}: {:?}", entry.block, e);
+					continue;
+				}
+			};
+			if cid.hash().code() != u64::from(cid::multihash::Code::Blake2b256)
+				|| cid.hash().size() != 32
+			{
+				debug!(target: LOG_TARGET, "Ignoring unsupported CID {}: {}", peer, cid);
+				continue
+			}
+			let mut hash = B::Hash::default();
+			hash.as_mut().copy_from_slice(&cid.hash().digest()[0..32]);
+			let extrinsic = match self.client.extrinsic(&hash) {
+				Ok(ex) => ex,
+				Err(e) => {
+					error!(target: LOG_TARGET, "Error retrieving extrinsic {}: {}", hash, e);
+					None
+				}
+			};
+			match extrinsic {
+				Some(extrinsic) => {
+					trace!(target: LOG_TARGET, "Found CID {:?}, hash {:?}", cid, hash);
+					if entry.want_type == WantType::Block as i32 {
+						let prefix = Prefix {
+							version: cid.version(),
```

Comment:
We could assert version is version 1 here (would be better to have a way to write prefix from cid crate but I didn't find it there either).

Reply:
bitswap 1.1.0 implies CIDv1 apparently. I'll add a check anyway